tools/pythonpkg/src/vector_conversion.cpp

```diff
@@ -0,0 +1,290 @@
+			}
+			tgt_ptr[row] = Timestamp::FromEpochNanoSeconds(src_ptr[source_idx]);
+		}
+		break;
+	}
+	case PandasType::VARCHAR: {
+		auto src_ptr = (PyObject **)numpy_col.data();
+		auto tgt_ptr = FlatVector::GetData<string_t>(out);
+		for (idx_t row = 0; row < count; row++) {
+			auto source_idx = offset + row;
+			auto val = src_ptr[source_idx];
+#if PY_MAJOR_VERSION >= 3
+			// Python 3 string representation:
+			// https://github.com/python/cpython/blob/3a8fdb28794b2f19f6c8464378fb8b46bce1f5f4/Include/cpython/unicodeobject.h#L79
+			if (!PyUnicode_CheckExact(val)) {
+				FlatVector::SetNull(out, row, true);
+				continue;
+			}
+			if (PyUnicode_IS_COMPACT_ASCII(val)) {
+				// ascii string: we can zero copy
+				tgt_ptr[row] = string_t((const char *)PyUnicode_DATA(val), PyUnicode_GET_LENGTH(val));
+			} else {
+				// unicode gunk
+				auto ascii_obj = (PyASCIIObject *)val;
+				auto unicode_obj = (PyCompactUnicodeObject *)val;
+				// compact unicode string: is there utf8 data available?
+				if (unicode_obj->utf8) {
+					// there is! zero copy
+					tgt_ptr[row] = string_t((const char *)unicode_obj->utf8, unicode_obj->utf8_length);
+				} else if (PyUnicode_IS_COMPACT(unicode_obj) && !PyUnicode_IS_ASCII(unicode_obj)) {
+					auto kind = PyUnicode_KIND(val);
+					switch (kind) {
+					case PyUnicode_1BYTE_KIND:
+						tgt_ptr[row] =
+						    DecodePythonUnicode<Py_UCS1>(PyUnicode_1BYTE_DATA(val), PyUnicode_GET_LENGTH(val), out);
+						break;
+					case PyUnicode_2BYTE_KIND:
+						tgt_ptr[row] =
+						    DecodePythonUnicode<Py_UCS2>(PyUnicode_2BYTE_DATA(val), PyUnicode_GET_LENGTH(val), out);
+						break;
+					case PyUnicode_4BYTE_KIND:
+						tgt_ptr[row] =
+						    DecodePythonUnicode<Py_UCS4>(PyUnicode_4BYTE_DATA(val), PyUnicode_GET_LENGTH(val), out);
+						break;
+					default:
+						throw std::runtime_error("Unsupported typekind for Python Unicode Compact decode");
+					}
+				} else if (ascii_obj->state.kind == PyUnicode_WCHAR_KIND) {
+					throw std::runtime_error("Unsupported: decode not ready legacy string");
+				} else if (!PyUnicode_IS_COMPACT(unicode_obj) && ascii_obj->state.kind != PyUnicode_WCHAR_KIND) {
+					throw std::runtime_error("Unsupported: decode ready legacy string");
+				} else {
+					throw std::runtime_error("Unsupported string type: no clue what this string is");
+				}
+			}
+#else
+			if (PyString_CheckExact(val)) {
+				auto dataptr = PyString_AS_STRING(val);
+				auto size = PyString_GET_SIZE(val);
+				// string object: directly pass the data
+				if (Utf8Proc::Analyze(dataptr, size) == UnicodeType::INVALID) {
+					throw std::runtime_error("String contains invalid UTF8! Please encode as UTF8 first");
+				}
+				tgt_ptr[row] = string_t(dataptr, uint32_t(size));
+			} else if (PyUnicode_CheckExact(val)) {
+				throw std::runtime_error("Unicode is only supported in Python 3 and up.");
+			} else {
+				FlatVector::SetNull(out, row, true);
+				continue;
+			}
+#endif
+		}
+		break;
+	}
+	default:
```

Comment:
There are couple more common numpy datatypes, one is timedelta and another one is object. Timedelta is basically just an integer, while object is a fall-back data types and it's probably fine to raise on it.